direct/nn/unet/unet_3d.py

```diff
@@ -0,0 +1,439 @@
+            ch *= 2
+        self.conv = ConvBlock3D(ch, ch * 2, dropout_probability)
+
+        self.up_conv = nn.ModuleList()
+        self.up_transpose_conv = nn.ModuleList()
+        for _ in range(num_pool_layers - 1):
+            self.up_transpose_conv += [TransposeConvBlock3D(ch * 2, ch)]
+            self.up_conv += [ConvBlock3D(ch * 2, ch, dropout_probability)]
+            ch //= 2
+
+        self.up_transpose_conv += [TransposeConvBlock3D(ch * 2, ch)]
+        self.up_conv += [
+            nn.Sequential(
+                ConvBlock3D(ch * 2, ch, dropout_probability),
+                nn.Conv3d(ch, out_channels, kernel_size=1, stride=1),
+            )
+        ]
+
+    def forward(self, input_data: torch.Tensor) -> torch.Tensor:
+        """Performs forward pass of :class:`UnetModel3d`.
+
+        Parameters
+        ----------
+        input_data : torch.Tensor
+            Input tensor of shape (N, in_channels, slice/time, height, width).
+
+        Returns
+        -------
+        torch.Tensor
+            Output of shape (N, out_channels, slice/time, height, width).
+        """
+        stack = []
+        output, inp_pad = pad_to_pow_of_2(input_data, self.num_pool_layers)
+
+        # Apply down-sampling layers
+        for _, layer in enumerate(self.down_sample_layers):
+            output = layer(output)
+            stack.append(output)
+            output = F.avg_pool3d(output, kernel_size=2, stride=2, padding=0)
+
+        output = self.conv(output)
+
+        # Apply up-sampling layers
+        for transpose_conv, conv in zip(self.up_transpose_conv, self.up_conv):
+            downsample_layer = stack.pop()
+            output = transpose_conv(output)
+
+            padding = [0, 0, 0, 0, 0, 0]
+            if output.shape[-1] != downsample_layer.shape[-1]:
+                padding[1] = 1
+            if output.shape[-2] != downsample_layer.shape[-2]:
+                padding[3] = 1
+            if output.shape[-3] != downsample_layer.shape[-3]:
+                padding[5] = 1
+            if sum(padding) != 0:
+                output = F.pad(output, padding, "reflect")
+
+            output = torch.cat([output, downsample_layer], dim=1)
+            output = conv(output)
+
+        if sum(inp_pad) != 0:
+            output = output[
+                :,
+                :,
+                inp_pad[4] : output.shape[2] - inp_pad[5],
+                inp_pad[2] : output.shape[3] - inp_pad[3],
+                inp_pad[0] : output.shape[4] - inp_pad[1],
+            ]
+
+        return output
+
+
+class NormUnetModel3d(nn.Module):
+    """Implementation of a Normalized U-Net model for 3D data."""
+
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        num_filters: int,
+        num_pool_layers: int,
+        dropout_probability: float,
+        norm_groups: int = 2,
+    ):
+        """Inits :class:`NormUnetModel3D`.
+
+        Parameters
```

Comment:
Perhaps a reference to where it was used before?

Reply:
Not sure if used somewhere. Referenced to our documentation